### ISSUE TITLE
API suggestion

### ISSUE THREAD

**Issue by sholland1** [benjamin-hodgson/Eighty#1]
Instead of returning a `TagBuilder` with the underscore method, you may want to do something like this:
```c#
public delegate B ParamsFunc<A, B>(params A[] ps);

public static ParamsFunc<Html, Html> p(/* attr params */)
{
    /* attr stuff */
    return new ParamsFunc<Html, Html>(children => new Tag(
        "p", ImmutableArrayFactory.UnsafeFreeze(array), children, false));
}
```

Usage would look like this:
```c#
var html = article(@class: "readme")( // no "._"
    h1(id: "Eighty")("Eighty"), // no "._"
    div(id: "no-children")(),
    hr(id: "self-closing"),
    p_(
        "Eighty (as in ",
        i_("eigh-ty-M-L"),
        ") is a simple HTML generation library."
    )
);
```

This will get rid of the `._` between the attributes and children. To maintain the overloads for the different collections, it looks like you would have to play with the names, which may not be worth the trade off.

I also wanted to mention that I couldn't get the underscore method to work from VB, since the underscore character is used for line continuations.

**Comment by benjamin-hodgson** [comment]
Thank you very much for your suggestion! I really appreciate your interest in my work and the effort involved in getting in touch. I'm always keen to hear how people are getting on with my libraries.

I did try out a `Func`-based API when I was first sketching out the idea for the library. There are a few reasons I decided against it:

* The `_` method is overloaded. You can pass 0-8`Html`s directly on the stack without having to build an array for the `params[]` overload. There are some other overloads to accept `IEnumerable`s etc too. You can't do this with a delegate
* It's wasteful to allocate a delegate only for it to immediately become garbage. (`TagBuilder` is a struct.)
* The `_` symbol is used quite consistently within the lib: it always introduces the children. `p()._(...)` and `p_(...)` mean the same thing. Changing `._(...)` to `(...)` would break that mnemonic.

That said, if C# ever adds some sort of "overloaded delegates" feature like Scala's `apply`, I would definitely reconsider this for a V2 of the library.

Regarding VB: I'm sorry you're having trouble getting it working. Eighty isn't CLS-compliant and it probably never will be - it was just never a goal of mine. I think your two best options are either to write your own CLS-compliant wrapping library (you'll need to use codegen to write all the HTML methods), or to have a go with Twenty, which I expect probably is CLS-compliant. Let me know how you get on!

**Comment by sholland1** [comment]
All valid points.

Actually, a couple of months ago I started working on a library like this after seeing the Lucid library from Haskell. I looked to see if anyone else had tried this kind of thing in .NET, but I didn't find anything worthwhile. The best thing I found was an [F# library](https://github.com/ptrelford/FsHtml), but that wouldn't work well outside of F#. I didn't get much done on the library and moved to other things.

This week I stumbled upon Eighty by chance. This encouraged me to play around with these ideas some more. I have uploaded what I have here: [HtmlDsl](https://github.com/sholland1/HtmlDsl). Feel free to have look.
